{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"14ZszkATtvSCYm7KJc21IGJ-oCrzHHxCx","timestamp":1732439131702}]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["!pip install scikit-surprise"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"orp5p-4IDBFP","executionInfo":{"status":"ok","timestamp":1732439257818,"user_tz":480,"elapsed":97727,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"8e3cd7f6-ba59-4975-b981-0668d0acf30e"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting scikit-surprise\n","  Downloading scikit_surprise-1.1.4.tar.gz (154 kB)\n","\u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/154.4 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m154.4/154.4 kB\u001b[0m \u001b[31m7.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n","  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n","  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n","Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from scikit-surprise) (1.4.2)\n","Requirement already satisfied: numpy>=1.19.5 in /usr/local/lib/python3.10/dist-packages (from scikit-surprise) (1.26.4)\n","Requirement already satisfied: scipy>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from scikit-surprise) (1.13.1)\n","Building wheels for collected packages: scikit-surprise\n","  Building wheel for scikit-surprise (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.4-cp310-cp310-linux_x86_64.whl size=2357288 sha256=92664b46560f4cfafdccdb1ba2ae2e56748de0ba44db5c4ed93ba0864365237d\n","  Stored in directory: /root/.cache/pip/wheels/4b/3f/df/6acbf0a40397d9bf3ff97f582cc22fb9ce66adde75bc71fd54\n","Successfully built scikit-surprise\n","Installing collected packages: scikit-surprise\n","Successfully installed scikit-surprise-1.1.4\n"]}]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0CYW6FaBF1Qg","executionInfo":{"status":"ok","timestamp":1732445016127,"user_tz":480,"elapsed":2205,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"231c80d6-cd4e-4c3a-e4b9-d6573a515bca"},"execution_count":20,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"code","execution_count":21,"metadata":{"id":"fwyqvSkLbgLh","executionInfo":{"status":"ok","timestamp":1732445020168,"user_tz":480,"elapsed":541,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}}},"outputs":[],"source":["import pandas as pd\n","from sklearn.model_selection import train_test_split\n","from sklearn.preprocessing import StandardScaler\n","from sklearn.linear_model import LogisticRegression\n","from sklearn.neighbors import KNeighborsClassifier\n","from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix\n","from sklearn.linear_model import LogisticRegression\n","from sklearn.metrics import mean_squared_error\n","from sklearn.ensemble import RandomForestClassifier\n","\n","from surprise import Dataset, Reader\n","from surprise import KNNWithMeans, SVD, BaselineOnly, accuracy\n","from surprise.model_selection import train_test_split\n","\n","import numpy as np"]},{"cell_type":"markdown","source":["# Multinomical Classification"],"metadata":{"id":"kd8Su2RygoRy"}},{"cell_type":"code","source":["train_data = pd.read_csv('/content/drive/MyDrive/CS 229 Group Project/Data/Updated/train_data(non-zero).csv')\n","test_data = pd.read_csv('/content/drive/MyDrive/CS 229 Group Project/Data/Updated/test_data(non-zero).csv')"],"metadata":{"id":"lxHdJwxJgshh","executionInfo":{"status":"ok","timestamp":1732445033532,"user_tz":480,"elapsed":2086,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}}},"execution_count":22,"outputs":[]},{"cell_type":"code","source":["# get number of unique users and unique products in each dataset\n","num_users_train = train_data['user_id'].nunique()\n","num_products_train = train_data['parent_asin'].nunique()\n","print(f\"Number of unique users in train data: {num_users_train}\")\n","print(f\"Number of unique products in train data: {num_products_train}\")\n","\n","num_users_test = test_data['user_id'].nunique()\n","num_products_test = test_data['parent_asin'].nunique()\n","print(f\"Number of unique users in train data: {num_users_test}\")\n","print(f\"Number of unique products in train data: {num_products_test}\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mBaMfZfthKkv","executionInfo":{"status":"ok","timestamp":1732445037318,"user_tz":480,"elapsed":316,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"ff7c94ed-6f48-49cf-8386-0f9d8f430d7c"},"execution_count":23,"outputs":[{"output_type":"stream","name":"stdout","text":["Number of unique users in train data: 3245\n","Number of unique products in train data: 378\n","Number of unique users in train data: 4130\n","Number of unique products in train data: 661\n"]}]},{"cell_type":"code","source":["train_data.info()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"UsAMYI23hzkR","executionInfo":{"status":"ok","timestamp":1732445043834,"user_tz":480,"elapsed":338,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"39ab04de-b4e5-4464-b93d-525f14c75a92"},"execution_count":24,"outputs":[{"output_type":"stream","name":"stdout","text":["<class 'pandas.core.frame.DataFrame'>\n","RangeIndex: 3431 entries, 0 to 3430\n","Data columns (total 14 columns):\n"," #   Column                    Non-Null Count  Dtype  \n","---  ------                    --------------  -----  \n"," 0   user_id                   3431 non-null   object \n"," 1   parent_asin               3431 non-null   object \n"," 2   rating                    3431 non-null   float64\n"," 3   user_average_rating       3431 non-null   float64\n"," 4   user_std_rating           3431 non-null   float64\n"," 5   avg_user_sentiment_title  3431 non-null   float64\n"," 6   avg_user_sentiment_text   3431 non-null   float64\n"," 7   user_review_count         3431 non-null   int64  \n"," 8   helpfulness_ratio         3431 non-null   float64\n"," 9   average_rating            3431 non-null   float64\n"," 10  rating_number             3431 non-null   int64  \n"," 11  title_word_count          3431 non-null   int64  \n"," 12  weighted_rating           3431 non-null   float64\n"," 13  description_sentiment     3431 non-null   float64\n","dtypes: float64(9), int64(3), object(2)\n","memory usage: 375.4+ KB\n"]}]},{"cell_type":"code","source":["user_features = [\n","    'user_average_rating', 'user_std_rating', 'avg_user_sentiment_title', 'avg_user_sentiment_text','user_review_count', 'helpfulness_ratio'\n","]\n","\n","item_columns = ['average_rating', 'rating_number', 'title_word_count', 'weighted_rating', 'description_sentiment']"],"metadata":{"id":"WgyAex1phoY6","executionInfo":{"status":"ok","timestamp":1732445054574,"user_tz":480,"elapsed":586,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}}},"execution_count":25,"outputs":[]},{"cell_type":"code","source":["train_data['rating'] = train_data['rating'].astype(int)\n","test_data['rating'] = test_data['rating'].astype(int)"],"metadata":{"id":"DOXUR1eJjXZB","executionInfo":{"status":"ok","timestamp":1732445162684,"user_tz":480,"elapsed":341,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}}},"execution_count":28,"outputs":[]},{"cell_type":"code","source":["# Assuming X_test is a pandas DataFrame\n","nan_rows = test_data[test_data.isnull().any(axis=1)]\n","\n","# Display the rows with NaN values\n","print(nan_rows)\n","\n","# To get the indices of the rows with NaN values:\n","nan_indices = nan_rows.index\n","print(nan_indices)\n","\n","# drop these\n","test_data = test_data.dropna()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ij4EvrMqx1Hc","executionInfo":{"status":"ok","timestamp":1732445212492,"user_tz":480,"elapsed":733,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"509e0d3e-4fa0-4d41-cdb4-e7f38c2d4a02"},"execution_count":30,"outputs":[{"output_type":"stream","name":"stdout","text":["                           user_id parent_asin  rating  user_average_rating  \\\n","3445  AEAZU2CWMBHHJJA4VHMXBRRH5QOA  B08C2GDFYG       4             4.333333   \n","3446  AEBKQHXFWNNJSRJZ7DRARIOUAGRA  B08C2GDFYG       5             5.000000   \n","3447  AEFOKABSQH2FQGUTODBNSG44B6YQ  B08C2GDFYG       5             5.000000   \n","\n","      user_std_rating  avg_user_sentiment_title  avg_user_sentiment_text  \\\n","3445          1.21106                  0.246138                 0.185479   \n","3446          0.00000                  0.250000                 0.330556   \n","3447          0.00000                  0.425000                -0.012500   \n","\n","      user_review_count  helpfulness_ratio  average_rating  rating_number  \\\n","3445                  6                7.5             4.5            761   \n","3446                  2                0.0             4.5            761   \n","3447                  2                0.0             4.5            761   \n","\n","      title_word_count  weighted_rating  description_sentiment  \n","3445               NaN         29.86176                    0.0  \n","3446               NaN         29.86176                    0.0  \n","3447               NaN         29.86176                    0.0  \n","Index([3445, 3446, 3447], dtype='int64')\n"]}]},{"cell_type":"code","source":["X_train = train_data[user_features + item_columns]\n","y_train = train_data['rating']\n","X_test = test_data[user_features + item_columns]\n","y_test = test_data['rating']\n","\n","# Scaling features\n","scaler = StandardScaler()\n","X_train_scaled = scaler.fit_transform(X_train)\n","X_test_scaled = scaler.transform(X_test)\n","\n","# Multinomial logistic regression\n","log_reg = LogisticRegression(max_iter=1000, class_weight='balanced', multi_class='multinomial', solver='lbfgs')\n","log_reg.fit(X_train_scaled, y_train)\n","\n","# Predictions\n","y_train_pred = log_reg.predict(X_train_scaled)\n","y_test_pred = log_reg.predict(X_test_scaled)\n","\n","# Metrics\n","train_accuracy = accuracy_score(y_train, y_train_pred)\n","test_accuracy = accuracy_score(y_test, y_test_pred)\n","\n","print(\"Training Accuracy:\", train_accuracy)\n","print(\"Testing Accuracy:\", test_accuracy)\n","\n","from sklearn.metrics import classification_report\n","print(classification_report(y_test, y_test_pred))\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"WLa3BYqpikyh","executionInfo":{"status":"ok","timestamp":1732445214869,"user_tz":480,"elapsed":252,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"e59ecd3a-0ac3-41e0-df0b-9e41404d9732"},"execution_count":31,"outputs":[{"output_type":"stream","name":"stdout","text":["Training Accuracy: 0.6674438939084815\n","Testing Accuracy: 0.6820137693631669\n","              precision    recall  f1-score   support\n","\n","           1       0.41      0.94      0.58       413\n","           2       0.27      0.29      0.28       256\n","           3       0.49      0.26      0.34       404\n","           4       0.49      0.43      0.46       625\n","           5       0.87      0.79      0.83      2950\n","\n","    accuracy                           0.68      4648\n","   macro avg       0.51      0.54      0.50      4648\n","weighted avg       0.71      0.68      0.68      4648\n","\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.\n","  warnings.warn(\n"]}]},{"cell_type":"markdown","source":["# Random Forest"],"metadata":{"id":"GtTDvzHgj-31"}},{"cell_type":"code","source":["from sklearn.ensemble import RandomForestClassifier\n","rf = RandomForestClassifier(n_estimators=100, random_state=42, class_weight='balanced')\n","rf.fit(X_train_scaled, y_train)\n","\n","# Predictions\n","y_train_pred = rf.predict(X_train_scaled)\n","y_test_pred = rf.predict(X_test_scaled)\n","\n","# Metrics\n","train_accuracy = accuracy_score(y_train, y_train_pred)\n","test_accuracy = accuracy_score(y_test, y_test_pred)\n","\n","print(\"Training Accuracy:\", train_accuracy)\n","print(\"Testing Accuracy:\", test_accuracy)\n","\n","print(classification_report(y_test, y_test_pred))\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"k417iy_6kBym","executionInfo":{"status":"ok","timestamp":1732445221164,"user_tz":480,"elapsed":1298,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"43a78366-6988-4d87-c014-5ada83aca489"},"execution_count":32,"outputs":[{"output_type":"stream","name":"stdout","text":["Training Accuracy: 1.0\n","Testing Accuracy: 0.7312822719449226\n","              precision    recall  f1-score   support\n","\n","           1       0.58      0.47      0.52       413\n","           2       0.57      0.18      0.28       256\n","           3       0.47      0.15      0.22       404\n","           4       0.61      0.39      0.47       625\n","           5       0.77      0.97      0.86      2950\n","\n","    accuracy                           0.73      4648\n","   macro avg       0.60      0.43      0.47      4648\n","weighted avg       0.69      0.73      0.69      4648\n","\n"]}]},{"cell_type":"markdown","source":["# SVM"],"metadata":{"id":"WsTx5dy3k6fH"}},{"cell_type":"code","source":["from sklearn.svm import SVC\n","svm = SVC(kernel='rbf', class_weight='balanced', probability=True)\n","svm.fit(X_train_scaled, y_train)\n","\n","# Predictions\n","y_train_pred = svm.predict(X_train_scaled)\n","y_test_pred = svm.predict(X_test_scaled)\n","\n","# Metrics\n","train_accuracy = accuracy_score(y_train, y_train_pred)\n","test_accuracy = accuracy_score(y_test, y_test_pred)\n","\n","print(\"Training Accuracy:\", train_accuracy)\n","print(\"Testing Accuracy:\", test_accuracy)\n","\n","print(classification_report(y_test, y_test_pred))\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"W-UE16Ovk8ro","executionInfo":{"status":"ok","timestamp":1732445232827,"user_tz":480,"elapsed":7435,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"dcc5b89b-e741-4449-ff31-075bdb87017e"},"execution_count":33,"outputs":[{"output_type":"stream","name":"stdout","text":["Training Accuracy: 0.6499562809676479\n","Testing Accuracy: 0.6536144578313253\n","              precision    recall  f1-score   support\n","\n","           1       0.47      0.86      0.61       413\n","           2       0.27      0.53      0.36       256\n","           3       0.37      0.49      0.42       404\n","           4       0.49      0.50      0.49       625\n","           5       0.93      0.69      0.79      2950\n","\n","    accuracy                           0.65      4648\n","   macro avg       0.50      0.61      0.53      4648\n","weighted avg       0.74      0.65      0.68      4648\n","\n"]}]},{"cell_type":"markdown","source":["# XGBoost"],"metadata":{"id":"796JvzEVlF47"}},{"cell_type":"code","source":["from xgboost import XGBClassifier\n","xgb = XGBClassifier(objective='multi:softprob', num_class=5, eval_metric='mlogloss')\n","xgb.fit(X_train_scaled, y_train-1)\n","\n","# Predictions\n","y_train_pred = xgb.predict(X_train_scaled)\n","y_test_pred = xgb.predict(X_test_scaled)\n","\n","# Metrics\n","train_accuracy = accuracy_score(y_train, y_train_pred)\n","test_accuracy = accuracy_score(y_test, y_test_pred)\n","\n","print(\"Training Accuracy:\", train_accuracy)\n","print(\"Testing Accuracy:\", test_accuracy)\n","\n","print(classification_report(y_test, y_test_pred))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"EG3_4PeSlIa-","executionInfo":{"status":"ok","timestamp":1732445240518,"user_tz":480,"elapsed":757,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"e7094791-9068-4a98-f9a7-c4b069d34a3a"},"execution_count":34,"outputs":[{"output_type":"stream","name":"stdout","text":["Training Accuracy: 0.0017487612940833576\n","Testing Accuracy: 0.07809810671256455\n","              precision    recall  f1-score   support\n","\n","           0       0.00      0.00      0.00         0\n","           1       0.07      0.02      0.04       413\n","           2       0.14      0.13      0.14       256\n","           3       0.13      0.19      0.15       404\n","           4       0.07      0.39      0.12       625\n","           5       0.00      0.00      0.00      2950\n","\n","    accuracy                           0.08      4648\n","   macro avg       0.07      0.12      0.08      4648\n","weighted avg       0.04      0.08      0.04      4648\n","\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n","  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n","/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n","  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n","/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n","  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n","/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n","  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n","/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n","  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n","/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.\n","  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"]}]},{"cell_type":"markdown","source":["# NN"],"metadata":{"id":"73qMV4seldlI"}},{"cell_type":"code","source":["import tensorflow as tf\n","model = tf.keras.models.Sequential([\n","    tf.keras.layers.Dense(128, activation='relu', input_shape=(X_train_scaled.shape[1],)),\n","    tf.keras.layers.Dropout(0.5),\n","    tf.keras.layers.Dense(64, activation='relu'),\n","    tf.keras.layers.Dense(5, activation='softmax')\n","])\n","model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])\n","model.fit(X_train_scaled, y_train, epochs=10, validation_split=0.2)\n","\n","test_loss, test_accuracy = model.evaluate(X_test_scaled, y_test)\n","\n","print(f\"Test Accuracy: {test_accuracy:.2f}\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":1000},"id":"vCEVc-4AleYD","executionInfo":{"status":"error","timestamp":1732445246569,"user_tz":480,"elapsed":1691,"user":{"displayName":"Sue Deng","userId":"17883311022612336035"}},"outputId":"765d5850-ff5e-4daf-afda-25658a1b2248"},"execution_count":35,"outputs":[{"output_type":"stream","name":"stdout","text":["Epoch 1/10\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.\n","  super().__init__(activity_regularizer=activity_regularizer, **kwargs)\n"]},{"output_type":"error","ename":"InvalidArgumentError","evalue":"Graph execution error:\n\nDetected at node compile_loss/sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits defined at (most recent call last):\n  File \"/usr/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n\n  File \"/usr/lib/python3.10/runpy.py\", line 86, in _run_code\n\n  File \"/usr/local/lib/python3.10/dist-packages/colab_kernel_launcher.py\", line 37, in <module>\n\n  File \"/usr/local/lib/python3.10/dist-packages/traitlets/config/application.py\", line 992, in launch_instance\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelapp.py\", line 619, in start\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/platform/asyncio.py\", line 195, in start\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n\n  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 685, in <lambda>\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 738, in _run_callback\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 825, in inner\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 786, in run\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 361, in process_one\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 261, in dispatch_shell\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 539, in execute_request\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py\", line 302, in do_execute\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/zmqshell.py\", line 539, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2975, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3030, in _run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3257, in run_cell_async\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n\n  File \"<ipython-input-35-d359293298be>\", line 9, in <cell line: 9>\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 117, in error_handler\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 320, in fit\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 121, in one_step_on_iterator\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 108, in one_step_on_data\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 54, in train_step\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/trainer.py\", line 359, in _compute_loss\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/trainer.py\", line 327, in compute_loss\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/compile_utils.py\", line 611, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/compile_utils.py\", line 652, in call\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses/loss.py\", line 60, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses/losses.py\", line 27, in call\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses/losses.py\", line 1870, in sparse_categorical_crossentropy\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/ops/nn.py\", line 1559, in sparse_categorical_crossentropy\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/nn.py\", line 671, in sparse_categorical_crossentropy\n\nReceived a label value of 5 which is outside the valid range of [0, 5).  Label values: 5 1 5 5 5 5 5 5 4 3 1 3 4 5 5 4 5 3 4 5 5 5 1 5 5 5 5 5 5 5 5 5\n\t [[{{node compile_loss/sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits}}]] [Op:__inference_one_step_on_iterator_2764]","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mInvalidArgumentError\u001b[0m                      Traceback (most recent call last)","\u001b[0;32m<ipython-input-35-d359293298be>\u001b[0m in \u001b[0;36m<cell line: 9>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m ])\n\u001b[1;32m      8\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moptimizer\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'adam'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloss\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'sparse_categorical_crossentropy'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmetrics\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'accuracy'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train_scaled\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalidation_split\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mtest_loss\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest_accuracy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mevaluate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_test_scaled\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    120\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m             \u001b[0;31m# `keras.config.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 122\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    123\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tensorflow/python/eager/execute.py\u001b[0m in \u001b[0;36mquick_execute\u001b[0;34m(op_name, num_outputs, inputs, attrs, ctx, name)\u001b[0m\n\u001b[1;32m     51\u001b[0m   \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m     \u001b[0mctx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mensure_initialized\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,\n\u001b[0m\u001b[1;32m     54\u001b[0m                                         inputs, attrs, num_outputs)\n\u001b[1;32m     55\u001b[0m   \u001b[0;32mexcept\u001b[0m \u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_NotOkStatusException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mInvalidArgumentError\u001b[0m: Graph execution error:\n\nDetected at node compile_loss/sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits defined at (most recent call last):\n  File \"/usr/lib/python3.10/runpy.py\", line 196, in _run_module_as_main\n\n  File \"/usr/lib/python3.10/runpy.py\", line 86, in _run_code\n\n  File \"/usr/local/lib/python3.10/dist-packages/colab_kernel_launcher.py\", line 37, in <module>\n\n  File \"/usr/local/lib/python3.10/dist-packages/traitlets/config/application.py\", line 992, in launch_instance\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelapp.py\", line 619, in start\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/platform/asyncio.py\", line 195, in start\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 603, in run_forever\n\n  File \"/usr/lib/python3.10/asyncio/base_events.py\", line 1909, in _run_once\n\n  File \"/usr/lib/python3.10/asyncio/events.py\", line 80, in _run\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 685, in <lambda>\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py\", line 738, in _run_callback\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 825, in inner\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 786, in run\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 361, in process_one\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 261, in dispatch_shell\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\", line 539, in execute_request\n\n  File \"/usr/local/lib/python3.10/dist-packages/tornado/gen.py\", line 234, in wrapper\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py\", line 302, in do_execute\n\n  File \"/usr/local/lib/python3.10/dist-packages/ipykernel/zmqshell.py\", line 539, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 2975, in run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3030, in _run_cell\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py\", line 78, in _pseudo_sync_runner\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3257, in run_cell_async\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3473, in run_ast_nodes\n\n  File \"/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\", line 3553, in run_code\n\n  File \"<ipython-input-35-d359293298be>\", line 9, in <cell line: 9>\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/utils/traceback_utils.py\", line 117, in error_handler\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 320, in fit\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 121, in one_step_on_iterator\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 108, in one_step_on_data\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/trainer.py\", line 54, in train_step\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/trainer.py\", line 359, in _compute_loss\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/trainer.py\", line 327, in compute_loss\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/compile_utils.py\", line 611, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/trainers/compile_utils.py\", line 652, in call\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses/loss.py\", line 60, in __call__\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses/losses.py\", line 27, in call\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/losses/losses.py\", line 1870, in sparse_categorical_crossentropy\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/ops/nn.py\", line 1559, in sparse_categorical_crossentropy\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/src/backend/tensorflow/nn.py\", line 671, in sparse_categorical_crossentropy\n\nReceived a label value of 5 which is outside the valid range of [0, 5).  Label values: 5 1 5 5 5 5 5 5 4 3 1 3 4 5 5 4 5 3 4 5 5 5 1 5 5 5 5 5 5 5 5 5\n\t [[{{node compile_loss/sparse_categorical_crossentropy/SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits}}]] [Op:__inference_one_step_on_iterator_2764]"]}]}]}